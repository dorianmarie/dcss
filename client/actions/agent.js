import {
  GET_AGENT_ERROR,
  GET_AGENT_SUCCESS,
  GET_AGENTS_ERROR,
  GET_AGENTS_SUCCESS,
  SET_AGENT_ERROR,
  SET_AGENT_SUCCESS
} from './types';

export let getAgent = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/agents/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const { agent } = res;

    dispatch({ type: GET_AGENT_SUCCESS, agent });
    return agent;
  } catch (error) {
    dispatch({ type: GET_AGENT_ERROR, error });
    return null;
  }
};

export let setAgent = (id, params) => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/agents/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { agent } = res;
      dispatch({ type: SET_AGENT_SUCCESS, agent });
      return agent;
    }
    return null;
  } catch (error) {
    dispatch({ type: SET_AGENT_ERROR, error });
    return null;
  }
};

export let createAgent = params => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/agents`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { agent } = res;
      dispatch({ type: GET_AGENT_SUCCESS, agent });
      return agent;
    }
    return null;
  } catch (error) {
    dispatch({ type: GET_AGENT_ERROR, error });
    return null;
  }
};

export let getAgents = filter => async dispatch => {
  try {
    const url =
      filter === 'is_active' ? '/api/agents/is_active' : '/api/agents';

    const res = await (await fetch(url)).json();

    if (res.error) {
      throw res;
    }

    const { agents = [] } = res;
    dispatch({ type: GET_AGENTS_SUCCESS, agents });
    return agents;
  } catch (error) {
    dispatch({ type: GET_AGENTS_ERROR, error });
    return null;
  }
};
