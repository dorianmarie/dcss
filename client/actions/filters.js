import { SET_FILTER_SCENARIOS_IN_USE } from './types';

export let setFilterScenariosInUse = scenariosInUse => dispatch => {
  dispatch({ type: SET_FILTER_SCENARIOS_IN_USE, scenariosInUse });
};
