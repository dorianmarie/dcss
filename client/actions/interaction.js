import {
  GET_INTERACTION_ERROR,
  GET_INTERACTION_SUCCESS,
  GET_INTERACTIONS_ERROR,
  GET_INTERACTIONS_SUCCESS,
  SET_INTERACTION_ERROR,
  SET_INTERACTION_SUCCESS,
GET_INTERACTIONS_TYPES_SUCCESS,
GET_INTERACTIONS_TYPES_ERROR,
} from './types';

export let getInteraction = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/interactions/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const { interaction } = res;

    dispatch({ type: GET_INTERACTION_SUCCESS, interaction });
    return interaction;
  } catch (error) {
    dispatch({ type: GET_INTERACTION_ERROR, error });
    return null;
  }
};

export let setInteraction = (id, params) => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/interactions/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { interaction } = res;
      dispatch({ type: SET_INTERACTION_SUCCESS, interaction });
      return interaction;
    }
    return null;
  } catch (error) {
    dispatch({ type: SET_INTERACTION_ERROR, error });
    return null;
  }
};

export let createInteraction = params => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/interactions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { interaction } = res;
      dispatch({ type: GET_INTERACTION_SUCCESS, interaction });
      return interaction;
    }
    return null;
  } catch (error) {
    dispatch({ type: GET_INTERACTION_ERROR, error });
    return null;
  }
};

export let getInteractions = () => async dispatch => {
  try {
    const res = await (await fetch(`/api/interactions`)).json();

    if (res.error) {
      throw res;
    }
    const { interactions } = res;

    dispatch({ type: GET_INTERACTIONS_SUCCESS, interactions });
    return interactions;
  } catch (error) {
    dispatch({ type: GET_INTERACTIONS_ERROR, error });
    return null;
  }
};

export let getInteractionsTypes = () => async dispatch => {
  try {
    const res = await (await fetch(`/api/interactions/types`)).json();

    if (res.error) {
      throw res;
    }
    const { types } = res;

    dispatch({ type: GET_INTERACTIONS_TYPES_SUCCESS, types });
    return types;
  } catch (error) {
    dispatch({ type: GET_INTERACTIONS_TYPES_ERROR, error });
    return null;
  }
};
