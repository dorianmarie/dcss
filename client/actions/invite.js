import {
  GET_INVITES_ERROR,
  GET_INVITES_SUCCESS,
  SET_INVITE_ERROR,
  SET_INVITE_SUCCESS
} from './types';

export let getInvites = () => async dispatch => {
  try {
    const res = await (await fetch(`/api/invites`)).json();

    if (res.error) {
      throw res;
    }
    const { invites } = res;
    dispatch({ type: GET_INVITES_SUCCESS, invites });
    return invites;
  } catch (error) {
    dispatch({ type: GET_INVITES_ERROR, error });
    return null;
  }
};

export let setInvite = (id, params) => async dispatch => {
  try {
    const { status } = params;
    const updates = {};

    if (status) {
      updates.status = status;
    }

    const body = JSON.stringify(updates);
    const res = await (await fetch(`/api/invites/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();

    if (res.error) {
      throw res;
    }

    const { invites } = res;
    const invite = invites.find(invite => invite.id === id);

    dispatch({ type: SET_INVITE_SUCCESS, invite });
    return invite;
  } catch (error) {
    dispatch({ type: SET_INVITE_ERROR, error });
    return null;
  }
};
