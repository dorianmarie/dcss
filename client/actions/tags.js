import {
  GET_CATEGORIES_ERROR,
  GET_CATEGORIES_SUCCESS,
  GET_LABELS_ERROR,
  GET_LABELS_SUCCESS,
  // NOT YET IN USE
  // CREATE_TAG_ERROR,
  // CREATE_TAG_SUCCESS,
  // SET_LABELS_IN_USE_ERROR,
  SET_LABELS_IN_USE_SUCCESS
} from './types';

export let getCategories = () => async dispatch => {
  try {
    const res = await (await fetch('/api/tags/categories')).json();

    if (res.error) {
      throw res;
    }

    const { categories } = res;

    dispatch({ type: GET_CATEGORIES_SUCCESS, categories });
    return categories;
  } catch (error) {
    dispatch({ type: GET_CATEGORIES_ERROR, error });
    return null;
  }
};

export let getLabels = () => async dispatch => {
  try {
    const res = await (await fetch('/api/tags/labels')).json();

    if (res.error) {
      throw res;
    }

    const { labels } = res;

    dispatch({ type: GET_LABELS_SUCCESS, labels });
    return labels;
  } catch (error) {
    dispatch({ type: GET_LABELS_ERROR, error });
    return null;
  }
};

export let getLabelsByOccurrence = (direction = 'desc') => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/tags/labels/occurrence/${direction}`
    )).json();

    if (res.error) {
      throw res;
    }

    const { labels } = res;

    dispatch({ type: GET_LABELS_SUCCESS, labels });
    return labels;
  } catch (error) {
    dispatch({ type: GET_LABELS_ERROR, error });
    return null;
  }
};

// This is intentionally NOT ASYNC
export let setLabelsInUse = labelsInUse => dispatch => {
  dispatch({ type: SET_LABELS_IN_USE_SUCCESS, labelsInUse });
};

/*
export let createTag = params => async dispatch => {

  // params {
  //   name
  //   tag_type_id
  //   (scenario_id)
  // }

  try {
    const res = await (
      await fetch('/api/tags', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(params)
      })
    ).json();

    if (res.error) {
      throw res;
    }
    const { tag } = res;
    dispatch({ type: CREATE_TAG_SUCCESS, tag });
    return tag;
  } catch (error) {
    dispatch({ type: CREATE_TAG_ERROR, error });
    return null;
  }
};
*/
