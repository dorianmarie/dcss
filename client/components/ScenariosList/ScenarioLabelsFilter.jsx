import React from 'react';
import PropTypes from 'prop-types';
import escapeRegExp from 'lodash.escaperegexp';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Dropdown, Text } from '@components/UI';
import { setLabelsInUse } from '@actions/tags';
import Identity from '@utils/Identity';
import QueryString from '@utils/QueryString';

function makeHistoryEntry(location, keyVals) {
  return `${location.pathname}?${QueryString.mergedStringify(keyVals)}`;
}

class ScenarioLabelsFilter extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      search: ''
    };

    this.onClick = this.onClick.bind(this);
  }

  onClick(event, { value }) {
    event.stopPropagation();

    const labelsInUse = this.props.tags.labelsInUse.slice(0);
    const labelsInUseBefore = labelsInUse.slice(0);

    if (labelsInUse.includes(value)) {
      labelsInUse.splice(labelsInUse.indexOf(value), 1);
    } else {
      labelsInUse.push(value);
    }

    if (this.props.onChange) {
      this.props.onChange(labelsInUseBefore, labelsInUse);
    }

    this.props.setLabelsInUse(labelsInUse);
    this.props.history.push(
      makeHistoryEntry(this.props.location, { l: labelsInUse })
    );
  }

  render() {
    const { search } = this.state;
    const { onClick } = this;
    const labels = [];

    // Move selected labels to the top of the list
    if (this.props.tags.labelsInUse.length) {
      const labelsSlice = this.props.tags.labels.slice();
      const head = [];
      const tail = [];

      while (labelsSlice.length) {
        const label = labelsSlice.shift();
        if (this.props.tags.labelsInUse.includes(label.value)) {
          head.push(label);
        } else {
          tail.push(label);
        }
      }
      labels.push(...head, ...tail);
    } else {
      labels.push(...(this.props.tags.labels || []));
    }

    let escapedRegExp;

    if (search) {
      escapedRegExp = new RegExp(escapeRegExp(search), 'i');
    }

    return (
      <Dropdown
        button
        item
        search
        scrolling
        text="Filter by labels"
        closeOnChange={false}
        onSearchChange={(e, props) => {
          this.setState({
            search: props.searchQuery
          });
        }}
      >
        <Dropdown.Menu className="labels">
          {labels.reduce((accum, label) => {
            const key = Identity.key(label);
            const active = this.props.tags.labelsInUse.includes(label.value);
            const content = (
              <div className="labels__item">
                <p>{label.text} </p>
                <Text size="small" grey>
                  {label.count} {label.count > 1 ? 'scenarios' : 'scenario'}
                </Text>
              </div>
            );
            const { text, value } = label;

            const itemProps = {
              'aria-label': `${label.count} scenarios labelled "${label.text}"`,
              active,
              content,
              key,
              onClick,
              text,
              value
            };

            if (active) {
              accum.push(<Dropdown.Item {...itemProps} />);
            } else {
              let shouldIncludeInDisplay = true;

              if (
                search &&
                !escapedRegExp.test(label.text) &&
                !escapedRegExp.test(label.value)
              ) {
                shouldIncludeInDisplay = false;
              }

              if (shouldIncludeInDisplay) {
                accum.push(<Dropdown.Item {...itemProps} />);
              }
            }
            return accum;
          }, [])}
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

ScenarioLabelsFilter.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  tags: PropTypes.shape({
    labels: PropTypes.array,
    labelsInUse: PropTypes.array
  }),
  location: PropTypes.object,
  setLabelsInUse: PropTypes.func,
  onChange: PropTypes.func
};

const mapStateToProps = state => {
  const { tags } = state;
  return { tags };
};

const mapDispatchToProps = dispatch => ({
  setLabelsInUse: params => dispatch(setLabelsInUse(params))
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ScenarioLabelsFilter)
);
