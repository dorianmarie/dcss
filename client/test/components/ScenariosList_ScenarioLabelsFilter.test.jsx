/** @TEMPLATE: BEGIN **/
import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
/** @TEMPLATE: END **/

/** @GENERATED: BEGIN **/

import * as tlr from '@testing-library/react';

import {
  GET_LABELS_SUCCESS,
  SET_LABELS_IN_USE_SUCCESS
} from '../../actions/types';
import * as tagsActions from '../../actions/tags';
jest.mock('../../actions/tags');

import * as QueryString from 'query-string';
jest.mock('query-string');

const labelA = { key: 1, text: 'a', value: 'a', count: 10 };
const labelB = { key: 2, text: 'b', value: 'b', count: 20 };
const labelC = { key: 3, text: 'c', value: 'c', count: 30 };

let labels;
let labelsByOccurrence;

import ScenarioLabelsFilter from '../../components/ScenariosList/ScenarioLabelsFilter.jsx';
/** @GENERATED: END **/

/** @TEMPLATE: BEGIN **/
const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;
/** @TEMPLATE: END **/

beforeAll(() => {
  /** @TEMPLATE: BEGIN **/
  (window || global).fetch = jest.fn();
  /** @TEMPLATE: END **/
});

afterAll(() => {
  /** @TEMPLATE: BEGIN **/
  jest.restoreAllMocks();
  /** @TEMPLATE: END **/
});

beforeEach(() => {
  /** @TEMPLATE: BEGIN **/
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
  /** @TEMPLATE: END **/

  /** @GENERATED: BEGIN **/

  labels = [labelA, labelB, labelC];

  labelsByOccurrence = [labelC, labelB, labelA];

  tagsActions.getLabels = jest.fn();

  tagsActions.getLabels.mockImplementation(() => async dispatch => {
    dispatch({ type: GET_LABELS_SUCCESS, labels });
    return labels;
  });
  tagsActions.getLabelsByOccurrence = jest.fn();

  tagsActions.getLabelsByOccurrence.mockImplementation(() => async dispatch => {
    const labels = labelsByOccurrence;
    dispatch({ type: GET_LABELS_SUCCESS, labels });
    return labels;
  });
  tagsActions.setLabelsInUse = jest.fn();

  tagsActions.setLabelsInUse.mockImplementation(labels => async dispatch => {
    dispatch({ type: SET_LABELS_IN_USE_SUCCESS, labels });
    return labels;
  });

  delete window.location;
  // eslint-disable-next-line
  window.location = {
    search: ''
  };

  QueryString.parse.mockImplementation(() => {
    return {
      l: ['a'],
      search: 'x'
    };
  });

  /** @GENERATED: END **/

  /** @TEMPLATE: BEGIN **/
  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
  /** @TEMPLATE: END **/
});

afterEach(() => {
  /** @TEMPLATE: BEGIN **/
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
  /** @TEMPLATE: END **/
});

test('ScenarioLabelsFilter', () => {
  expect(ScenarioLabelsFilter).toBeDefined();
});

/** @GENERATED: BEGIN **/
test('Render 1 1', async done => {
  const Component = ScenarioLabelsFilter;
  const props = {
    ...commonProps
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});
/** @GENERATED: END **/

/* INJECTION STARTS HERE */

test('Selected labelsInUse', async done => {
  const Component = ScenarioLabelsFilter;

  const props = {
    ...(commonProps || {})
  };

  const tags = {
    categories: [],
    labels: labelsByOccurrence,
    labelsInUse: ['a']
  };

  const state = {
    ...commonState,
    tags
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});

test('No selected labelsInUse', async done => {
  const Component = ScenarioLabelsFilter;

  QueryString.parse.mockImplementation(() => {
    return {
      l: [],
      search: 'x'
    };
  });

  const props = {
    ...(commonProps || {})
  };

  const tags = {
    categories: [],
    labels: labelsByOccurrence,
    labelsInUse: []
  };

  const state = {
    ...commonState,
    tags
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});

test('Add and remove a label', async done => {
  const Component = ScenarioLabelsFilter;

  const props = {
    ...(commonProps || {})
  };

  const tags = {
    categories: [],
    labels: labelsByOccurrence,
    labelsInUse: ['a']
  };

  const state = {
    ...commonState,
    tags
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  const dropdown = await screen.getByRole('combobox');
  expect(asFragment()).toMatchSnapshot();

  const input = await tlr.findAllByDisplayValue(dropdown, '');
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(dropdown);
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(input[0]);
  userEvent.type(input[0], 'a{enter}');
  expect(asFragment()).toMatchSnapshot();

  let options = await tlr.findAllByRole(dropdown, 'option');

  userEvent.click(options[0]);
  expect(asFragment()).toMatchSnapshot();
  expect(tagsActions.setLabelsInUse).toHaveBeenCalledTimes(1);

  // The options are now re-ordered, so query again...
  options = await tlr.findAllByRole(dropdown, 'option');

  userEvent.click(options[0]);
  expect(asFragment()).toMatchSnapshot();
  expect(tagsActions.setLabelsInUse).toHaveBeenCalledTimes(2);

  // The options are now re-ordered, so query again...
  options = await tlr.findAllByRole(dropdown, 'option');

  userEvent.click(options[0]);
  expect(asFragment()).toMatchSnapshot();
  expect(tagsActions.setLabelsInUse).toHaveBeenCalledTimes(3);

  done();
});
