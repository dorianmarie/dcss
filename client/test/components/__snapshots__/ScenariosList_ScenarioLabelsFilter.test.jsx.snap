// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add and remove a label 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="30 scenarios labelled \\"c\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            c 
          </p>
          <span
            class="ui grey small text"
          >
            30 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="20 scenarios labelled \\"b\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            b 
          </p>
          <span
            class="ui grey small text"
          >
            20 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 2`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="30 scenarios labelled \\"c\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            c 
          </p>
          <span
            class="ui grey small text"
          >
            30 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="20 scenarios labelled \\"b\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            b 
          </p>
          <span
            class="ui grey small text"
          >
            20 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 3`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="30 scenarios labelled \\"c\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            c 
          </p>
          <span
            class="ui grey small text"
          >
            30 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="20 scenarios labelled \\"b\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            b 
          </p>
          <span
            class="ui grey small text"
          >
            20 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 4`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    class="ui active visible button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition visible labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="30 scenarios labelled \\"c\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            c 
          </p>
          <span
            class="ui grey small text"
          >
            30 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="20 scenarios labelled \\"b\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            b 
          </p>
          <span
            class="ui grey small text"
          >
            20 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 5`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    class="ui active visible button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value="a"
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text filtered"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition visible labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 6`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    class="ui active visible button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value="a"
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text filtered"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition visible labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 7`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    class="ui active visible button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value="a"
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text filtered"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition visible labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Add and remove a label 8`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    class="ui active visible button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value="a"
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text filtered"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition visible labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`No selected labelsInUse 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition labels"
      role="listbox"
    >
      <div
        aria-checked="false"
        aria-label="30 scenarios labelled \\"c\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            c 
          </p>
          <span
            class="ui grey small text"
          >
            30 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="20 scenarios labelled \\"b\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            b 
          </p>
          <span
            class="ui grey small text"
          >
            20 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="10 scenarios labelled \\"a\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render 1 1 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="undefined scenarios labelled \\"Easy\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            Easy 
          </p>
          <span
            class="ui grey small text"
          >
             scenario
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="undefined scenarios labelled \\"fun\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            fun 
          </p>
          <span
            class="ui grey small text"
          >
             scenario
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Selected labelsInUse 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui button item search scrolling dropdown"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="search"
      tabindex="0"
      type="text"
      value=""
    />
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition labels"
      role="listbox"
    >
      <div
        aria-checked="true"
        aria-label="10 scenarios labelled \\"a\\""
        class="active item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            a 
          </p>
          <span
            class="ui grey small text"
          >
            10 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="30 scenarios labelled \\"c\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            c 
          </p>
          <span
            class="ui grey small text"
          >
            30 scenarios
          </span>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-label="20 scenarios labelled \\"b\\""
        class="item"
        role="option"
      >
        <div
          class="text labels__item"
        >
          <p>
            b 
          </p>
          <span
            class="ui grey small text"
          >
            20 scenarios
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
